libs/grok/src/main/java/org/elasticsearch/grok/GrokCaptureConfig.java

```diff
@@ -50,18 +54,104 @@ public String name() {
     /**
      * The type defined for the field in the pattern.
      */
-    public GrokCaptureType type() {
+    GrokCaptureType type() {
         return type;
     }
 
-    Object extract(byte[] utf8Bytes, int offset, Region region) {
-        for (int number : backRefs) {
-            if (region.beg[number] >= 0) {
-                String matchValue = new String(utf8Bytes, offset + region.beg[number], region.end[number] - region.beg[number],
-                    StandardCharsets.UTF_8);
-                return type.parse(matchValue);
+    /**
+     * Build a {@linkplain GrokCaptureExtracter} that will call {@code emit} when
+     * it extracts text, boxed if the "native" representation is primitive type.
+     * Extracters returned from this method are stateless and can be reused.
+     */
+    GrokCaptureExtracter objectExtracter(Consumer<Object> emit) {
+        return nativeExtracter(new NativeExtracterMap<GrokCaptureExtracter>() {
+            @Override
+            public GrokCaptureExtracter forString(Function<Consumer<String>, GrokCaptureExtracter> buildExtracter) {
+                return buildExtracter.apply(str -> emit.accept(str));
+            }
+
+            @Override
+            public GrokCaptureExtracter forInt(Function<IntConsumer, GrokCaptureExtracter> buildExtracter) {
+                return buildExtracter.apply(i -> emit.accept(Integer.valueOf(i)));
```

Comment:
This particular one (and really all the other native type parsers), are passed the value of `X.parseX`. So, it is actually passed the result of `Integer.parseInt(/*String*/i)`. It then follows that boxing here is not necessary. I think you can just keep it native all the way as the `IntConsumer` is designed to accept native values.

Reply:
```suggestion
                return buildExtracter.apply(emit::accept);
```
Given the above comment, I think this is all that is necessary (and this goes for all the other extractor methods too).

Reply:
That doesn't so much keep it native as rely on autoboxing. I intentionally didn't use autoboxing here because I figured it'd be clearer that we *have* to box here to get the `Object`.

Reply:
I see, definitely, since the resulting Object must be returned to the ultimate consumer. 

Stinks that we lose the native values and that they must be boxed.

Reply:
Yeah! This is mostly for the folks that need it wrapped in an object anyway like ingest. Runtime fields will use the native values.